qiskit_experiments/curve_analysis/base_curve_analysis.py

```diff
@@ -0,0 +1,480 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Base class of curve analysis.
+"""
+
+import warnings
+
+from abc import ABC, abstractmethod
+from typing import List, Dict, Union, Optional
+
+from uncertainties import unumpy as unp
+
+from qiskit_experiments.framework import BaseAnalysis, AnalysisResultData, Options, ExperimentData
+from qiskit_experiments.data_processing import DataProcessor
+from qiskit_experiments.exceptions import AnalysisError
+
+from .curve_data import CurveData, SeriesDef, FitData, ParameterRepr, FitOptions
+from .curve_fit import multi_curve_fit
+from .visualization import MplCurveDrawer, BaseCurveDrawer
+
+PARAMS_ENTRY_PREFIX = "@Parameters_"
+DATA_ENTRY_PREFIX = "@Data_"
+
+
+class BaseCurveAnalysis(BaseAnalysis, ABC):
+    """Abstract superclass of curve analysis base classes.
+
+    Note that this class doesn't define :meth:`_run_analysis` method,
+    and no actual fitting protocol is implemented by itself.
+    However, this class defines several common methods that can be reused.
+    A curve analysis subclass can construct proper fitting protocol
+    by combining following methods, i.e. sub-routines.
+
+    _generate_fit_guesses
+
+        An abstract method to create initial guees.
+        This should be implemented by subclass.
+
+    _format_data
+
+        A method to format curve data. By default this method takes y value average
+        over the same x values and then sort the entire data by x values.
```

Comment:
No this is always done. Because some guess function relies on the sequence ordering, e.g. computing slope by something like `dx = x[1] - x[0]`. There is no guarantee x values are in ascending order.

Reply:
There are two things here.

a) averaging the y data
b) sorting according to x data.

Makes sense to always do b). Do we always do a)?

Reply:
Yes, should do. If you concatenates two experiments, you will create experiment data that may consist of something like `x = [0,1,2,3,0,1,2,3]` that will crash some guess function.